### PR TITLE
feat(pipeline): implement normalize progress tracking for JSON normalizer

### PR DESCRIPTION
**Pull request by IlyaFaer**
Towards #1567

Implements item 1 for `JsonLItemsNormalizer`

### CODE REVIEW

File: dlt/normalize/items_normalizers.py [hunk 1]
```diff
@@ -83,9 +93,15 @@ def _normalize_chunk(
         schema_name = schema.name
         normalize_data_fun = self.schema.normalize_data_item
 
+        total = self.extract_info.total_rows_count if self.extract_info else None
+
+        import time
+
         for item in items:
+            time.sleep(0.7)
```
Comment (IlyaFaer) [inline dlt/normalize/items_normalizers.py:101]: I left his testing stuff for now, so you'll be able to see that it works

File: dlt/normalize/items_normalizers.py [hunk 2]
```diff
@@ -83,9 +93,15 @@ def _normalize_chunk(
...
             items_gen = normalize_data_fun(item, self.load_id, root_table_name)
             try:
+                self.collector.update("Items", total=total)
```
Comment (IlyaFaer) [inline dlt/normalize/items_normalizers.py:104]: For every row we update the collector

**Comment by IlyaFaer** [comment]
For the case when we're first extracting data and then normalizing, it works fine

image

What I don't like is that we have to through collector and extract data through a bunch of methods and classes, but for now I don't see any solution, as the code is chained.

Thinking about how to test it... Seems to me something cunning is needed.

File: dlt/normalize/items_normalizers.py [hunk 3]
```diff
@@ -83,9 +87,13 @@ def _normalize_chunk(
         schema_name = schema.name
         normalize_data_fun = self.schema.normalize_data_item
 
+        # import time
+
         for item in items:
+            # time.sleep(0.7)
```
Comment (IlyaFaer) [inline dlt/normalize/items_normalizers.py:93]: Use it for testing

**Comment by IlyaFaer** [comment]
@rudolfix, hm-m, looks like pipeline state is a better solution - allows to cover both cases, when the data was just extracted, and when it's been extracted on the previous run.

File: dlt/pipeline/pipeline.py [hunk 4]
```diff
@@ -445,7 +445,12 @@ def extract(
                 )
                 # commit load packages with state
                 extract_step.commit_packages()
-                return self._get_step_info(extract_step)
+
+                info = self._get_step_info(extract_step)
+                state = self._container[StateInjectableContext].state
+                state["_local"]["_last_extracted_count"] = info.total_rows_count
```
Comment (IlyaFaer) [inline dlt/pipeline/pipeline.py:451]: Figuring out if it's possible to do this in a less hacky way without breaking anything

**Comment by rudolfix** [comment]
collected metrics will not go through process boundary, we'll need different approach here

**Pull request closed without merging**